### PR TITLE
Preserve ordering equivalencies on `with_reorder`

### LINKED ISSUE #13769: Preserve ordering equivalencies on `with_reorder`
### Is your feature request related to a problem or challenge?

When applying new orderings via `with_reorder()`, the current implementation discards all existing ordering information and replaces it with the new ordering. This is suboptimal because in many cases, parts of the existing ordering remain valid and could be preserved.

For example, if we have data ordered by `[a ASC, b DESC]` and apply a new ordering `[a ASC]`, we currently lose the information that the data is also ordered by `b DESC`, even though this ordering is still valid and could be valuable for query optimization.

This overly conservative approach can lead to unnecessary sorting operations in query execution, impacting query performance.

### Describe the solution you'd like

Enhance `with_reorder()` to preserve valid ordering information by:

1. Filtering out constant expressions from ordering considerations (they don't affect physical ordering)
2. Preserving valid suffixes from existing orderings when they're compatible with the new ordering
3. Considering expression equivalences when comparing orderings

For example:
- When applying `[a ASC]` to existing `[a ASC, b DESC]`, preserve `b DESC`
- When columns `a` and `b` are equivalent and applying `[b ASC]` to existing `[a ASC, c DESC]`, preserve `c DESC`
- When column `a` is constant, applying `[a ASC, b ASC]` should only consider `[b ASC]`

### Describe alternatives you've considered

Only preserve exact matching prefixes without considering equivalences

### Additional context

This optimization is particularly important for complex queries with multiple ordering requirements and transformations, where maintaining ordering information can significantly reduce the number of required sort operations.

### PR DESCRIPTION
**Pull request by gokselk**
## Which issue does this PR close?

Closes #13769.

## Rationale for this change

The current implementation of `with_reorder` discards all existing ordering information when applying a new ordering. This is overly conservative since in many cases, parts of the existing ordering remain valid and could be preserved. Preserving valid orderings is important for query optimization as it allows the optimizer to avoid unnecessary sorting operations.

## What changes are included in this PR?

1. Enhanced `with_reorder()` to preserve valid suffixes from existing orderings when they don't conflict with the new ordering
2. Added filtering of constant expressions from ordering considerations
3. Added `exprs_equal()` helper method to `EquivalenceGroup` for comparing expressions considering equivalence classes
4. Added comprehensive test suite covering:
   - Constant expression filtering
   - Suffix preservation
   - Equivalent expression handling
   - Incompatible prefix cases
   - Complex scenarios combining multiple features

## Are these changes tested?

Yes, the PR includes extensive test coverage:
- `test_with_reorder_constant_filtering`: Tests filtering of constant expressions
- `test_with_reorder_preserve_suffix`: Verifies preservation of valid ordering suffixes
- `test_with_reorder_equivalent_expressions`: Tests handling of equivalent expressions
- `test_with_reorder_incompatible_prefix`: Ensures proper handling of incompatible orderings
- `test_with_reorder_comprehensive`: Tests complex interactions between constants, equivalences, and multiple orderings

## Are there any user-facing changes?

This change is an internal optimization that preserves more ordering information. While there are no direct API changes, users may notice improved query performance in scenarios where ordering information is better preserved through operations that modify sort orders.

### CODE REVIEW

**Comment by gokselk** [comment]
cc: @berkaysynnada @ozankabak

**Review by berkaysynnada — APPROVED** [review]
LGTM

File: datafusion/physical-expr/src/equivalence/class.rs [hunk 1]
```diff
@@ -634,6 +634,18 @@ impl EquivalenceGroup {
             JoinType::RightSemi | JoinType::RightAnti => right_equivalences.clone(),
         }
     }
+
+    /// Checks if two expressions are equal either directly or through equivalence classes.
+    pub fn exprs_equal(
```
Comment (ozankabak) [inline datafusion/physical-expr/src/equivalence/class.rs:639]: This function does not work for two general expressions. In general, checking whether two expression trees are equal involves a lot of work: One not only needs to make sure all symbols (i.e. columns) are equivalent, but also take operator properties like commutativity into account.

For the purposes of this PR, I think it is sufficient to just make sure that:
1. `left` and `right` expression trees are "the same" except leaves.
2. Each leaf on the left and right side are equivalent w.r.t. some group.

Comment (ozankabak) [reply datafusion/physical-expr/src/equivalence/class.rs:639]: An easy way to test would be to look at `a + b` and `x + y` with `a` and `x` / `b` and `y` being equivalent.

**Review by ozankabak — COMMENTED** [review]
I have one comment, looks good after we address it (and add a test)

**Comment by alamb** [comment]
Is this PR ready to merge?

I triggered the CI checks again

File: datafusion/physical-expr/src/equivalence/class.rs (hunk 1 above)
Comment (gokselk) [reply datafusion/physical-expr/src/equivalence/class.rs:641]: Resolved in 4d37673.

**Comment by berkaysynnada** [comment]
resolved the conflicts, will be merged once CI passes

**Pull request merged**